packages/userstorage/src/browser/user-storage-service-filesystem.ts

```diff
@@ -0,0 +1,92 @@
+import { FileSystem, FileChange } from '@theia/filesystem/lib/common';
+import { FileSystemWatcher } from "@theia/filesystem/lib/common";
+import { stripComments } from 'jsonc-parser';
+import URI from '@theia/core/lib/common/uri';
+import { UserStorageUri } from './user-storage-uri';
+
+export const THEIA_USER_STORAGE_FOLDER = '.theia';
+
+@injectable()
+export class UserStorageServiceFilesystemImpl implements UserStorageService {
+
+    protected readonly toDispose = new DisposableCollection();
+    protected readonly onUserStorageChangedEmitter = new Emitter<UserStorageChangeEvent>();
+    protected userStorageFolder: Promise<URI>;
+
+    constructor(
+        @inject(FileSystem) protected readonly fileSystem: FileSystem,
+        @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
+        @inject(ILogger) protected readonly logger: ILogger
+
+    ) {
+        this.userStorageFolder = this.fileSystem.getCurrentUserHome().then(home => {
+
+            const userStorageFolderUri = new URI(home.uri).resolve(THEIA_USER_STORAGE_FOLDER);
+            watcher.watchFileChanges(userStorageFolderUri).then(disposable =>
+                this.toDispose.push(disposable)
+            );
+            this.toDispose.push(this.watcher.onFilesChanged(changes => this.onDidFilesChanged(changes)));
+            return new URI(home.uri).resolve(THEIA_USER_STORAGE_FOLDER);
+        });
+    }
+
+    dispose(): void {
+        this.toDispose.dispose();
+    }
+
+    onDidFilesChanged(fileChanges: FileChange[]): void {
+        const uris: string[] = [];
+        for (const change of fileChanges) {
+            uris.push(this.fromFsUri(change.uri.toString()).toString());
+        }
+        this.onUserStorageChangedEmitter.fire({ uris });
+    }
+
+    readContents(uri: string) {
+        return this.userStorageFolder
+            .then(folderUri => {
+                const userStorageUri = new URI(uri);
+                return this.fileSystem.resolveContent(this.toFsUri(folderUri, userStorageUri).toString());
+            })
+            .then(({ stat, content }) => stripComments(content));
+    }
+
+    saveContents(uri: string, content: string) {
+
+        return this.userStorageFolder.then(folderUri => {
+            const fsUri = this.toFsUri(folderUri, new URI(uri));
+
+            this.fileSystem.getFileStat(fsUri.toString()).then(fileStat => {
+                this.fileSystem.setContent(fileStat, content).then(() => Promise.resolve());
+            });
+        });
+    }
+
+    get onUserStorageChanged(): Event<UserStorageChangeEvent> {
+        return this.onUserStorageChangedEmitter.event;
+    }
+
+    /**
+     * Creates a new user storage URI from the filesystem path argument.
+     * @param fsPath the filesystem path.
+     */
+    private fromFsUri(fsPath: string) {
+        const fsUri = new URI(fsPath);
+        return new URI('').withScheme(UserStorageUri.SCHEME).withPath(fsUri.path.base);
```

Comment:
Did you overlook my comment about the implementation, or am I wrong?
> Also I don't see how it translates e.g. 
> - `file:///Users/username/.theia/settings.json` to
> - `user_storage:///settings.json.`
> Seems like it would only replace the scheme, which is wrong, isn't it?

My point is, it would result in `user_storage:///Users/username/.theia/settings.json`.
Also please translate the query and fragment of the URI, too, if present.

Reply:
A test, that tests the symmetry of the two methods would help :)

Reply:
Ok I get your point, and I totally agree with the test cases for this, let me work on this.

Reply:
I changed the conversion a bit and added tests @svenefftinge, let me know what you think. I now keep the fragment and query as part of the uri structure now, even though I'm not really sure they can be useful for the fs?